src/traits/impls.rs

```diff
@@ -0,0 +1,132 @@
+//!
+//! Doughnut trait impls
+//!
+use super::{DoughnutApi, DoughnutVerify};
+use crate::error::ValidationError;
+#[cfg(feature = "std")]
+use ed25519_dalek::{PublicKey as Ed25519Pub, Signature as Ed25519Sig};
+#[cfg(feature = "std")]
+use schnorrkel::{signing_context, PublicKey as Sr25519Pub, Signature as Sr25519Sig};
+
+#[cfg(feature = "std")]
+impl<T: DoughnutApi> DoughnutVerify for T
+where
+    <T as DoughnutApi>::PublicKey: AsRef<[u8]>,
+    <T as DoughnutApi>::Signature: AsRef<[u8]>,
+{
+    fn verify(&self) -> bool {
+        match self.signature_version() {
+            0 => {
+                let signature = match Sr25519Sig::from_bytes(self.signature().as_ref()) {
+                    Ok(s) => s,
+                    Err(_) => return false,
+                };
+                match Sr25519Pub::from_bytes(self.issuer().as_ref()) {
+                    Ok(pk) => {
+                        // TODO: Parameterise signing context. Hard-coded for now
+                        pk.verify(
+                            signing_context(b"substrate").bytes(self.payload().as_ref()),
+                            &signature,
+                        )
+                    }
+                    Err(_) => return false,
+                }
+            }
+            1 => {
+                let signature = match Ed25519Sig::from_bytes(self.signature().as_ref()) {
+                    Ok(s) => s,
+                    Err(_) => return false,
+                };
+                match Ed25519Pub::from_bytes(self.issuer().as_ref()) {
+                    Ok(pk) => match pk.verify(&self.payload(), &signature) {
+                        Ok(_) => true,
+                        _ => false,
+                    },
+                    Err(_) => false,
+                }
+            }
+            _ => {
+                // Unsupported
+                false
+            }
+        }
+    }
+}
+
+// Dummy implementation for unit type
+impl DoughnutApi for () {
+    type PublicKey = ();
+    type Timestamp = ();
+    type Signature = ();
+    fn holder(&self) -> Self::PublicKey {
+        ()
+    }
+    fn issuer(&self) -> Self::PublicKey {
+        ()
+    }
+    fn expiry(&self) -> Self::PublicKey {
```

Comment:
Where is the `not_before` value?

Reply:
It's yet to be exposed.

Reply:
Ah, I think there's a misunderstanding here -`not_before` isn't a periodisation value!